modules/phase_field/src/postprocessors/GrainTracker.C

```diff
@@ -308,15 +308,30 @@ GrainTracker::trackGrains()
         mooseError("Error with ESBD Mapping (see above unused indices)");
       }
     }
-    else // Just assign IDs in order
+    else
     {
+      /**
+       * Here we want to assign the grains in some partitioning invariant way. We'll sort first by
+       * _var_idx (already partitioned in that manner) then sort on the _min_entity_id
+       * to order the grains.
+       */
       unsigned int counter = 0;
       for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
-        for (unsigned int feature_num = 0; feature_num < _feature_sets[map_num].size(); ++feature_num)
-        {
-          _feature_sets[map_num][feature_num]->_status = MARKED;
-          _unique_grains[counter++] = std::move(_feature_sets[map_num][feature_num]);
-        }
+      {
+        // Sort the grains represented by this variable by _min_entity_id
+        std::sort(_feature_sets[map_num].begin(), _feature_sets[map_num].end(),
+                  [](const std::unique_ptr<FeatureData> & lhs, const std::unique_ptr<FeatureData> & rhs)
+                  {
+                    return lhs->_min_entity_id < rhs->_min_entity_id;
+                  });
+
+        // Transfer the grains to the _unique_grains structure and assign ids
+        std::transform(_feature_sets[map_num].begin(), _feature_sets[map_num].end(), std::inserter(_unique_grains, _unique_grains.end()),
```

Comment:
I'm turning an array that looks like this { A, D, C, B, F, E}
into a map like this { (1,A), (2,D), (3,D), (4, B), (5, F), (6, E) }

Yeah this could be trivially turned into a `for` loop. Perhaps that would be better...